security/acme-client/src/opnsense/mvc/app/library/OPNsense/AcmeClient/LeUtils.php

```diff
@@ -49,6 +53,27 @@ public static function base64url_encode($str)
         return rtrim(strtr(base64_encode($str), '+/', '-_'), '=');
     }
 
+    /**
+     * Properly escape arguments to prevent shell command injection.
+     * This is a copy of the exec_safe() legacy function.
+     * @param string $format The format string to use.
+     * @param string $args The arguments to escape.
+     * @return array|string The formatted and escaped arguments.
+     */
+    public static function execSafe($format, $args = array())
+    {
+        if (!is_array($args)) {
+            /* just in case there's only one argument */
+            $args = array($args);
+        }
+
+        foreach ($args as $id => $arg) {
+            $args[$id] = escapeshellarg($arg);
+        }
+
+        return vsprintf($format, $args);
```

Comment:
@fraenki it's probably better to let "execSafe" do the execution as well, which would mean you need to reformat some blocks to feed `execSafe()` with all args, but would be more or less a drop in replacement for `exec_safe()` in our legacy code base. copying the functionality (with the correct authors :) ) is currently the best strategy to follow indeed.

Reply:
What do you mean by "let execSafe do the execution as well"? Should I modify the function to actually run the shell command?

Reply:
Shouldn’t pull in more emulation at the moment. I think Ad is talking long term.

Reply:
never mind, I totally misread the `exec_safe()` function internals, the name already was somewhat confusing it seems. It doesn't execute anything as the name suggests, merely cleanses the input (which is the same as `LeUtils::execSafe` is doing now).